### PR TITLE
[WPE][GTK] Drop libnotify

### PR DESCRIPTION
**Pull request by carlosgcampos**
#### 95858e6cc2e3f804cc95b4f55cd7532329db790a
<pre>
[WPE][GTK] Drop libnotify
<a href="https://bugs.webkit.org/show_bug.cgi?id=240347">https://bugs.webkit.org/show_bug.cgi?id=240347</a>

Reviewed by Michael Catanzaro.

Add NotificationService singleton to communicate with the notifications
service.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp:
(webkitNotificationCreate):
(webkitNotificationGetWebNotification):
(webkit_notification_get_id):
(webkit_notification_get_title):
(webkit_notification_get_body):
(webkit_notification_get_tag):
(webkitNotificationGetWebView): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitNotificationPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:
(WebKitNotificationProvider::~WebKitNotificationProvider):
(WebKitNotificationProvider::apiNotificationCloseCallback):
(WebKitNotificationProvider::apiNotificationClickedCallback):
(WebKitNotificationProvider::closeAPINotification):
(WebKitNotificationProvider::withdrawAnyPreviousAPINotificationMatchingTag):
(WebKitNotificationProvider::show):
(WebKitNotificationProvider::cancelNotificationByID):
(WebKitNotificationProvider::didClickNotification):
(WebKitNotificationProvider::didCloseNotification):
(WebKitNotificationProvider::notificationCloseCallback): Deleted.
(WebKitNotificationProvider::notificationClickedCallback): Deleted.
(WebKitNotificationProvider::withdrawAnyPreviousNotificationMatchingTag): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(notifyNotificationClosed): Deleted.
(notifyNotificationClicked): Deleted.
(webNotificationClosed): Deleted.
(webkitWebViewShowNotification): Deleted.
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp: Added.
(WebKit::NotificationService::singleton):
(WebKit::NotificationService::NotificationService):
(WebKit::NotificationService::processCapabilities):
(WebKit::NotificationService::showNotification):
(WebKit::NotificationService::cancelNotification):
(WebKit::NotificationService::setNotificationID):
(WebKit::NotificationService::findNotification):
(WebKit::NotificationService::handleSignal):
(WebKit::NotificationService::didClickNotification):
(WebKit::NotificationService::didCloseNotification):
(WebKit::NotificationService::addObserver):
(WebKit::NotificationService::removeObserver):
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.h: Added.
* Source/cmake/FindLibNotify.cmake: Removed.
* Source/cmake/OptionsGTK.cmake:

Canonical link: <a href="https://commits.webkit.org/253179@main">https://commits.webkit.org/253179@main</a>
</pre>

### CODE REVIEW

File: Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp [hunk 1]
```diff
@@ -51,12 +51,11 @@ enum {
 };
 
 struct _WebKitNotificationPrivate {
+    RefPtr<WebKit::WebNotification> notification;
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp:54]: Surely this should be:

```suggestion
    Ref<WebKit::WebNotification> notification;
```

because you're treating it as if it is not nullable everywhere.

File: Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp [hunk 2]
```diff
@@ -80,69 +81,95 @@ WebKitNotificationProvider::WebKitNotificationProvider(WebNotificationManagerPro
...
-    for (auto& notification : m_notifications.values()) {
+    for (auto& notification : m_apiNotifications.values()) {
         if (tag == webkit_notification_get_tag(notification.get())) {
-            webkit_notification_close(notification.get());
+            closeAPINotification(webkit_notification_get_id(notification.get()));
             break;
         }
     }
 
 #ifndef NDEBUG
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp]: Good opportunity for a drive-by fix:

```suggestion
#if ASSERT_ENABLED
```

(`NDEBUG` controls glibc assertions, not WebKit assertions.)

File: Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp [hunk 3]
```diff
@@ -80,69 +81,95 @@ WebKitNotificationProvider::WebKitNotificationProvider(WebNotificationManagerPro
...
 }
 
-void WebKitNotificationProvider::show(WebPageProxy* page, const WebNotification& webNotification)
+void WebKitNotificationProvider::show(WebPageProxy* page, WebNotification& webNotification)
 {
     if (!page) {
-        // FIXME: glib needs to find their own solution to handling pageless notifications.
-        // I won't make something up for them.
+        // FIXME: glib API needs to find their own solution to handling pageless notifications.
+        show(webNotification);
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:132]: So now we show the notification unconditionally, without allowing the application to control it? That's probably worse than dropping the notification, because it will bypass the application's permission store. Maybe the original behavior was better here?

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+static const Seconds s_dbusCallTimeout = 20_ms;
+
+NotificationService& NotificationService::singleton()
+{
+    static std::once_flag onceFlag;
+    static LazyNeverDestroyed<NotificationService> service;
+
+    std::call_once(onceFlag, [] {
+        service.construct();
+    });
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:55]: This class is not threadsafe, so you should just store a bool rather than using `std::call_once`. The point of `std::call_once` is to use a mutex to guarantee it gets initialized once regardless of thread.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    return service;
+}
+
+NotificationService::NotificationService()
+{
+    const char* busName = shouldUsePortal() ? "org.freedesktop.portal.Desktop" : "org.freedesktop.Notifications";
+    const char* objectPath = shouldUsePortal() ? "/org/freedesktop/portal/desktop" : "/org/freedesktop/Notifications";
+    const char* interfaceName = shouldUsePortal() ? "org.freedesktop.portal.Notification" : "org.freedesktop.Notifications";
+    GUniqueOutPtr<GError> error;
+    m_proxy = g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES, nullptr, busName, objectPath, interfaceName, nullptr, &error.outPtr());
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:66]: Why do you have to use sync I/O here? Wouldn't this be quite easy to convert to async?

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+    if (!m_proxy) {
+        g_warning("Failed to connect to notification service at %s: %s", busName, error->message);
+        return;
+    }
+
+    if (!shouldUsePortal()) {
+        GRefPtr<GVariant> capabilities = adoptGRef(g_dbus_proxy_call_sync(m_proxy.get(), "GetCapabilities", nullptr, G_DBUS_CALL_FLAGS_NONE, s_dbusCallTimeout.millisecondsAs<int>(), nullptr, &error.outPtr()));
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:73]: Ditto?

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+            m_capabilities.add(Capabilities::Sound);
+    }
+}
+
+bool NotificationService::showNotification(const WebNotification& notification)
+{
+    if (!m_proxy)
+        return false;
+
+    auto findNotificationByTag = [&](const String& tag) -> Notification {
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp]: Better to capture only what you need, rather than everything. I think the only thing you need here is `this`, right?

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    if (shouldUsePortal()) {
+        GVariantBuilder builder;
+        g_variant_builder_init(&builder, G_VARIANT_TYPE_VARDICT);
+
+        g_variant_builder_add(&builder, "{sv}", "title", g_variant_new_string(notification.title().utf8().data()));
+        g_variant_builder_add(&builder, "{sv}", "body", g_variant_new_string(notification.body().utf8().data()));
+        g_variant_builder_add(&builder, "{sv}", "default-action", g_variant_new_string("default"));
+        addResult.iterator->value.portalID = createVersion4UUIDString();
+        g_dbus_proxy_call(m_proxy.get(), "AddNotification", g_variant_new("(s@a{sv})", addResult.iterator->value.portalID.utf8().data(), g_variant_builder_end(&builder)),
+            G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp]: Should pass a callback just to print a warning if it fails. Yes, it's slightly more effort, but the future benefit in case things break in the future will be worth it.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    } else {
+        GVariantBuilder actionsBuilder;
+        g_variant_builder_init(&actionsBuilder, G_VARIANT_TYPE("as"));
+
+        if (m_capabilities.contains(Capabilities::Actions)) {
+            g_variant_builder_add(&actionsBuilder, "s", "default");
+            g_variant_builder_add(&actionsBuilder, "s", _("Acknowledge"));
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:159]:
```suggestion
            // TRANSLATORS: the default action for a desktop notification created by a website.
            g_variant_builder_add(&actionsBuilder, "s", _("Acknowledge"));
```

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+        const char* applicationID = nullptr;
+        if (auto* app = g_application_get_default())
+            applicationID = g_application_get_application_id(app);
+
+        GVariantBuilder hintsBuilder;
+        g_variant_builder_init(&hintsBuilder, G_VARIANT_TYPE("a{sv}"));
+        if (applicationID)
+            g_variant_builder_add(&hintsBuilder, "{sv}", "desktop-entry", g_variant_new_string(applicationID));
+
+        auto* value = static_cast<GValue*>(fastZeroedMalloc(sizeof(GValue)));
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:171]: Can you use `UniqueRef` for this?

If not, you can use a `std::unique_ptr` with a custom deleter and `WTFMove` it into the callback.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+
+    auto it = m_notifications.find(webNotificationID);
+    if (it == m_notifications.end())
+        return;
+
+    if (shouldUsePortal()) {
+        if (it->value.portalID.isEmpty())
+            return;
+
+        g_dbus_proxy_call(m_proxy.get(), "RemoveNotification", g_variant_new("(s)", it->value.portalID.utf8().data()), G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp]: I'd like to see callbacks everywhere you omit them currently. Even doing nothing except print error and return is valuable.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    } else {
+        if (!it->value.id)
+            return;
+
+        g_dbus_proxy_call(m_proxy.get(), "CloseNotification", g_variant_new("(u)", it->value.id), G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp]: Ditto.

**Review by mcatanzaro — APPROVED** [review]
Nice, thanks.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp (hunk 5 above)
Comment (TingPing) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:66]: Sync bus lookup is non-blocking after the first call. But the call below is blocking.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:66]: What's the "first call?"

File: Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp (hunk 1 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp:54]: That requires a value provided in the constructor and that's not possible with GObject priv struct.

File: Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp (hunk 3 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:132]: I don't think applications handle the permissions in show-notification callback, permissions are set in the context and handled by WebKit, and that's still working in this case, we don't even get here if permission hasn't been granted for the notification origin. We don't even expose the notification origin in the API.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp (hunk 4 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:55]: This is a common pattern in all singleton using LazyNeverDestroyed, I think it's harmless even if the class it not thread safe.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp (hunk 5 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:66]: Using async IO for these would make everything a lot more complicated for very little benefit. For the first notification show will happen before the proxy is constructed or capabilities processed, so we would need to handle the case and wait for the async to complete to show the notification. Since this only happens once I don't think it's worth it.

Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:66]: > What's the "first call?"

I think he means getting the bus connection, because g_bus_get_sync returns a singleton, so only the first time it's called it needs to connect which is async. At this point, glib has already connected to the session bus for sure.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp (hunk 8 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp]: Ok

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp (hunk 7 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp]: It's a lambda local to the function, but ok.

File: Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp (hunk 10 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:171]: I'll try

File: Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp (hunk 3 above)
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:132]: OK.
